docs: v1.1.0 (#419)
umijs/dumi: docs/guide/basic.md

---
title: Basic
order: 2
toc: menu
---

## Directory Structure

### Common project

> The directory structure of a common component library that uses dumi for research and development is roughly as follows:

Reviewer: ```suggestion
The directory structure of a common component library that based on dumi is roughly as follows:
```

<Tree>
  <ul>
    <li>
      docs
      <small>Component library document directory</small>
      <ul>
        <li>
          index.md

>           <small>Documentation Home (If it does not exist, it will fallback to README.md)</small>

Reviewer: ```suggestion
          <small>The homepage of documentation(If it does not exist, it will fallback to README.md)</small>
```

        </li>
        <li>
          guide

>           <small>Component library document other routing (Signal)</small>

Reviewer: ```suggestion
          <small>The other pages of documentation (Signal) </small>
```

          <ul>
            <li>index.md</li>
            <li>sample.md</li>
            <li>help.md</li>
          </ul>
        </li>
      </ul>
    </li>
    <li>
      src

>       <small>Source directory</small>

Reviewer: ```suggestion
      <small>Source code directory</small>
```

      <ul>
        <li>
          Button
          <small>Single component</small>
          <ul>
            <li>
              index.tsx

>               <small>Single component</small>

Reviewer: ```suggestion
              <small>The source code of this component</small>
```

            </li>
            <li>
              index.less

>               <small>Component style</small>

Reviewer: ```suggestion
              <small>The style of this component</small>
```

            </li>
            <li>
              index.md

>               <small>Component documentation</small>

Reviewer: ```suggestion
              <small>The documentation of this component</small>
```

            </li>
          </ul>
        </li>
        <li>
          index.ts

>           <small>Component library entry file</small>

Reviewer: ```suggestion
          <small>The entry file of the component library</small>
```

        </li>
      </ul>
    </li>
    <li>
      .umirc.ts
      <small>dumi configuration file (Can also be config/config.ts)</small>
    </li>
    <li>
      .fatherrc.ts
      <small>father-build configuration file, used for component library packaging</small>
    </li>
  </ul>
</Tree>

> If it is a pure document site without component source code, just ignore the above `src` directory structure.

Reviewer: ```suggestion
If it is a pure document site without component source code, just ignore the `src` directory above.
```

> Note that only the directory structure is explained here.

Reviewer: ```suggestion
Note that only explains what directory structure looks like here
```

If you want to initialize a dumi project, it is recommended to directly use the scaffolding of `@umijs/create-dumi-lib` or `@umijs/create-dumi-app` to create it.

### lerna project

In most cases, we will gather the documentation of all lerna sub-packages in the same documentation site.

The directory structure is usually like this:

<Tree>
  <ul>
    <li>
      docs
      <small>Public documents</small>
      <ul>
        <li>
          index.md

>           <small>Documentation Home (If it does not exist, it will fallback to README.md)</small>

Reviewer: ```suggestion
          <small>The homepage of documentation (If it does not exist, it will fallback to README.md)</small>
```

        </li>
        <li>
          guide

>           <small>Component library document other routing (Signal)</small>

Reviewer: ```suggestion
          <small>The other pages of documentation (Signal) </small>
```

As shown in the figure below, dumi's conventional routing rules are very simple:

<img src="https://gw.alipayobjects.com/zos/bmw-prod/66ca1c57-3146-40d5-bea4-24dea86eb95b/kijoyzad_w1540_h732.png" width="770" />

Give a few examples to facilitate understanding:

| Disk path/mode | doc mode | site mode |
| --- | --- | --- |
| /path/to/src/index.md | - group: none<br >- page routing: / | - navigate: none<br >- group: none<br>- page routing: / |
| /path/to/src/hello.md | - group: none<br >- page routing: /hello | - navigate: none<br >- group: none<br>- page routing: /hello |
| /path/to/src/hello/index.md | - group: /hello<br >- page routing: /hello | - navigate: /hello<br >- group:/hello<br>- page routing: /hello |
| /path/to/src/hello/world.md | - group: /hello<br >- page routing: /hello/world | - navigate: /hello<br >- group: /hello<br>- page routing: /hello/world |
| /path/to/src/hello/world/dumi.md | - group: /hello/world<br >- page routing: /hello/world/dumi | - navigate: /hello<br >- group: /hello/world<br>- page routing: /hello/world/dumi |

It should be noted that the routes generated by files with the same disk path under multiple base paths will conflict with each other.

> Which means that in the default configuration, `docs/index.md` and `src/index.md` are only among them 1 will be recognized.

Reviewer: ```suggestion
Which means that only one of  `docs/index.md` and `src/index.md` will be recognized in the default configuration.
```

### Customize navigation, grouping and titles

> If you want to control the generation of navigation/grouping/page title, you can do this by **writing FrontMatter at the top of the Markdown file**:

Reviewer: ```suggestion
If you want to control the generation of navigation/grouping/titles, you can control it by **writing FrontMatter at the top of the Markdown file**:
```

```markdown
---
title: Custom page name
nav:
  path: /Custom navigation route
  title: Custom navigation name
  order: Control the navigation order, the smaller the number, the higher the order, the default is to sort by path length and dictionary order
group:
  path: /Custom group routing, note that group routing = navigation routing + yourself
  title: Custom group name
  order: Control the grouping order, the smaller the number, the higher the order, the default is to sort by path length and dictionary order
---

<!-- Other Markdown content -->
```

In site mode, we can also incrementally customize the navigation and left menu through configuration items.

Please refer to [Configuration Item-navs]() and [Configuration Item-menus]().

## Write component demo

> dumi provides two ways to write demos to deal with different scenarios.

Reviewer: ```suggestion
Dumi provides two ways to write demos to deal with different scenarios.
```

### Code block

If our demo is very lightweight, it is recommended to write code blocks directly, such as:

<pre lang="markdown">
```jsx
import React from 'react';

export default () => <h1>Hello dumi!</h1>;
```
</pre>

The code blocks of `jsx` and `tsx` will be parsed as React components by dumi, and the above code blocks will be rendered as:

```jsx
import React from 'react';

export default () => <h1>Hello dumi!</h1>;
```

If we want a block of `jsx`/`tsx` code to be rendered as source code, we can use the `pure` modifier to tell dumi:

<pre lang="markdown">
```jsx | pure
// I will not be rendered as a React component
```
</pre>

### External demo

If our demo is very complicated and may even have many external files, then it is recommended to use an external demo:

```markdown
<code src="/path/to/complex-demo.tsx"></code>
```

> Like the code block demo, the above code will also be rendered as React components.

Reviewer: ```suggestion
Like the code block demo, the above code will also be rendered as a react component.
```
